tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -0,0 +1,190 @@
+// ./lesson_21_run
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_21_auto_scheduler_run
+// in a shell with the current directory at the top of the halide
+// source tree.
+
+#include "Halide.h"
+#include <stdio.h>
+
+using namespace Halide;
+
+// We will define a generator to auto-schedule.
+class AutoScheduleGen : public Halide::Generator<AutoScheduleGen> {
+public:
+    GeneratorParam<bool>  auto_schedule{"auto_schedule", false};
+
+    Input<Buffer<float>>  input{"input", 4};
+    Input<Buffer<float>>  filter{"filter", 4};
+    Input<Buffer<float>>  bias{"bias", 1};
+    Input<float>          min_value{"min_value"};
+
+    Output<Buffer<float>> output1{"output1", 4};
+    Output<Buffer<float>> output2{"output2", 4};
+
+    void generate() {
+        RDom r(filter.dim(0).min(), filter.dim(0).extent(),
+               filter.dim(1).min(), filter.dim(1).extent(),
+               filter.dim(2).min(), filter.dim(2).extent());
+
+        f(x, y, z, n) = bias(z);
+        f(x, y, z, n) += filter(r.x, r.y, r.z, z) * input(x + r.x, y + r.y, r.z, n);
+        output1(x, y, z, n) = max(0, f(x, y, z, n));
+        output2(x, y, z, n) = min(min_value, f(x, y, z, n));
+    }
+
+    void schedule() {
+        if (!auto_schedule) {
+            return;
+        }
+
+        // To use the auto-scheduler, we need to provide estimates on the
+        // input/output sizes including estimates on the parameter values.
+
+        // First, provide estimates (min and extent values) for each dimension
+        // of the input images ('input', 'filter', and 'bias') using the
+        // set_bounds_estimate() method. set_bounds_estimate() takes in
+        // (min, extent) of the corresponding dimension as arguments.
+        input.dim(0).set_bounds_estimate(0, 131);
+        input.dim(1).set_bounds_estimate(0, 131);
+        input.dim(2).set_bounds_estimate(0, 64);
+        input.dim(3).set_bounds_estimate(0, 4);
+
+        filter.dim(0).set_bounds_estimate(0, 3);
+        filter.dim(1).set_bounds_estimate(0, 3);
+        filter.dim(2).set_bounds_estimate(0, 64);
+        filter.dim(3).set_bounds_estimate(0, 64);
+
+        bias.dim(0).set_bounds_estimate(0, 64);
+
+        // Next, provide estimates on the parameter values using the
+        // set_estimate() method.
+        min_value.set_estimate(2.0f);
+
+        // Last, provide estimates (min and extent values) for each dimension
+        // of pipeline outputs using the estimate() method. estimate() takes in
+        // (dim_name, min, extent) as arguments.
+        output1.estimate(x, 0, 128)
+            .estimate(y, 0, 128)
+            .estimate(z, 0, 64)
+            .estimate(n, 0, 4);
+
+        output2.estimate(x, 0, 128)
+            .estimate(y, 0, 128)
+            .estimate(z, 0, 64)
+            .estimate(n, 0, 4);
+
+        // Technically, the estimate values can be anything, but the closer
```

Comment:
Probably worth adding a comment about what happens if you omit estimates for something?